sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/MessageBody.cs

```diff
@@ -143,23 +159,44 @@ public override IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
                 return _segments.GetEnumerator();
             }
 
-            private void Append(Data segment)
+            private void Append(IEnumerable<Data> dataSegments)
```

Comment:
The Append method is almost identical between CopyingOnConversionMessageBody and EagerCopyingMessageBody. Can we extract it somehow?

Reply:
Couldn't the code be exactly the same? The switch case would still work for the CopyOnConversion path, even though it is unnecessary. You could also just pass a flag.

Reply:
Regarding the coupling, I think that is where having tests for both paths helps, which I think we already have.

Reply:
I generally would not abstract things on such a path but even though I have different opinions I don't think I need to argue since I'm happy to adjust to what you as owners prefer.

How about something like this?

```csharp
static class MessageBodyHelper
    {
        public static void AppendSegments<T>(IEnumerable<T> dataSegments, Func<T, ReadOnlyMemory<byte>> converter, ref ArrayBufferWriter<byte> refWriter,
            ref IList<ReadOnlyMemory<byte>> refSegments)
        {
            int length = 0;
            int numberOfSegments = 0;
            List<ReadOnlyMemory<byte>> segments = null;
            foreach (var segment in dataSegments)
            {
                segments ??= dataSegments is IReadOnlyCollection<Data> readOnlyList
                    ? new List<ReadOnlyMemory<byte>>(readOnlyList.Count)
                    : new List<ReadOnlyMemory<byte>>();
                ReadOnlyMemory<byte> dataToAppend = converter(segment);
                length += dataToAppend.Length;
                numberOfSegments++;
                segments.Add(dataToAppend);
            }

            if (segments == null)
            {
                return;
            }

            // fields are lazy initialized to not occupy unnecessary memory when there are no data segments
            refWriter = length > 0 ? new ArrayBufferWriter<byte>(length) : new ArrayBufferWriter<byte>();
            refSegments = segments;

            for (var i = 0; i < numberOfSegments; i++)
            {
                var dataToAppend = segments[i];
                var memory = refWriter.GetMemory(dataToAppend.Length);
                dataToAppend.CopyTo(memory);
                refWriter.Advance(dataToAppend.Length);
                segments[i] = memory.Slice(0, dataToAppend.Length);
            }
        }
```

Reply:
That works for me!

Reply:
The price of the abstraction is non trivial. 

![image](https://user-images.githubusercontent.com/174258/191619903-a9b5a2e4-0a4e-4e7e-aa18-c92dc1e52997.png)

Reply:
Hmm what do you mean by abstraction? Do you mean calling the static helper method?

Reply:
Could it be due to passing the func? Could we instead just pass a flag or use the same logic for both cases?

Reply:
This is how it would look like

```csharp
       private sealed class CopyingOnConversionMessageBody : MessageBody
        {
            private ArrayBufferWriter<byte> _writer;
            private IList<ReadOnlyMemory<byte>> _segments;
            private IEnumerable<ReadOnlyMemory<byte>> _lazySegments;
            private static readonly Func<ReadOnlyMemory<byte>, ReadOnlyMemory<byte>> _mapToSelf = segment => segment;

            internal CopyingOnConversionMessageBody(IEnumerable<ReadOnlyMemory<byte>> dataSegments)
            {
                _lazySegments = dataSegments;
            }

            protected override ReadOnlyMemory<byte> WrittenMemory
            {
                get
                {
                    if (_lazySegments != null)
                    {
                        // fields are lazy initialized to not occupy unnecessary memory when there are no data segments
                        Segments.Append(_lazySegments,  _mapToSelf, ref _writer, ref _segments);
                        _lazySegments = null;
                    }

                    return _writer?.WrittenMemory ?? ReadOnlyMemory<byte>.Empty;
                }
            }

            public override IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
            {
                return _segments?.GetEnumerator() ?? _lazySegments.GetEnumerator();
            }
        }

        private sealed class EagerCopyingMessageBody : MessageBody
        {
            private readonly ArrayBufferWriter<byte> _writer;
            private readonly IList<ReadOnlyMemory<byte>> _segments;

            internal EagerCopyingMessageBody(IEnumerable<Data> dataSegments)
            {
                // fields are lazy initialized to not occupy unnecessary memory when there are no data segments
                Segments.Append(dataSegments,  static segment =>
                {
                    return segment.Value switch
                    {
                        byte[] byteArray => byteArray,
                        ArraySegment<byte> arraySegment => arraySegment,
                        _ => ReadOnlyMemory<byte>.Empty
                    };
                }, ref _writer, ref _segments);
            }

            protected override ReadOnlyMemory<byte> WrittenMemory => _writer?.WrittenMemory ?? ReadOnlyMemory<byte>.Empty;

            public override IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
            {
                return _segments.GetEnumerator();
            }
        }

        private static class Segments
        {
            public static void Append<T>(IEnumerable<T> dataSegments, Func<T, ReadOnlyMemory<byte>> converter, ref ArrayBufferWriter<byte> refWriter,
                ref IList<ReadOnlyMemory<byte>> refSegments)
            {
                int length = 0;
                int numberOfSegments = 0;
                List<ReadOnlyMemory<byte>> segments = null;
                foreach (var segment in dataSegments)
                {
                    segments ??= dataSegments is IReadOnlyCollection<Data> readOnlyList
                        ? new List<ReadOnlyMemory<byte>>(readOnlyList.Count)
                        : new List<ReadOnlyMemory<byte>>();
                    ReadOnlyMemory<byte> dataToAppend = converter(segment);
                    length += dataToAppend.Length;
                    numberOfSegments++;
                    segments.Add(dataToAppend);
                }

                if (segments == null)
                {
                    return;
                }

                refWriter = length > 0 ? new ArrayBufferWriter<byte>(length) : new ArrayBufferWriter<byte>();
                refSegments = segments;

                for (var i = 0; i < numberOfSegments; i++)
                {
                    var dataToAppend = segments[i];
                    var memory = refWriter.GetMemory(dataToAppend.Length);
                    dataToAppend.CopyTo(memory);
                    refWriter.Advance(dataToAppend.Length);
                    segments[i] = memory.Slice(0, dataToAppend.Length);
                }
            }
        }
```

Reply:
Another run of the current version

![image](https://user-images.githubusercontent.com/174258/191622435-4bd05072-ddc1-43b1-b809-9cd1fa93160d.png)

another run with the abstracted version

![image](https://user-images.githubusercontent.com/174258/191622704-905e1385-ae3b-4f93-bde6-1ca04128ee03.png)

I could try to run a very long run with a higher iteration count to see if I get less statistical deviation but these runs take ages :( Anyway I need to finish for today.

Reply:
@JoshLove-msft yes the helper method including the infrastructure it needs. That is what I mean as an abstraction.

> Could it be due to passing the func? Could we instead just pass a flag or use the same logic for both cases?

How would that work? One accepts `Data` segments and one `ReadOnlyMemory<byte>` there is no common ground to switch case over. I would need to do nested switches then 

```csharp
        private static class Segments
        {
            public static void Append<T>(IEnumerable<T> dataSegments, ref ArrayBufferWriter<byte> refWriter,
                ref IList<ReadOnlyMemory<byte>> refSegments)
            {
                int length = 0;
                int numberOfSegments = 0;
                List<ReadOnlyMemory<byte>> segments = null;
                foreach (var segment in dataSegments)
                {
                    segments ??= dataSegments is IReadOnlyCollection<Data> readOnlyList
                        ? new List<ReadOnlyMemory<byte>>(readOnlyList.Count)
                        : new List<ReadOnlyMemory<byte>>();
                    ReadOnlyMemory<byte> dataToAppend =  segment switch
                    {
                        ReadOnlyMemory<byte> romSegment => romSegment,
                        Data data => data.Value switch
                        {
                            byte[] byteArray => byteArray,
                            ArraySegment<byte> arraySegment => arraySegment,
                            _ => ReadOnlyMemory<byte>.Empty
                        },
                        _ => throw new ArgumentOutOfRangeException()
                    };
                    length += dataToAppend.Length;
                    numberOfSegments++;
                    segments.Add(dataToAppend);
                }

                if (segments == null)
                {
                    return;
                }

                refWriter = length > 0 ? new ArrayBufferWriter<byte>(length) : new ArrayBufferWriter<byte>();
                refSegments = segments;

                for (var i = 0; i < numberOfSegments; i++)
                {
                    var dataToAppend = segments[i];
                    var memory = refWriter.GetMemory(dataToAppend.Length);
                    dataToAppend.CopyTo(memory);
                    refWriter.Advance(dataToAppend.Length);
                    segments[i] = memory.Slice(0, dataToAppend.Length);
                }
            }
        }
```

Reply:
Pushed the new version. In case you decide you don't like it we can revert that commit

Reply:
I'm fine with the nested switch. @jsquire, what do you think?

Reply:
I'm good with either approach and defer to your judgement, @JoshLove-msft.

For my two cents, I don't mind a bit of code duplication when it is focused and serves a purpose.  In this case I prefer it; I found the original version to be simpler, clearer, and have better locality of information.  In particular, the reliance on mutation and side-effects in `Segments::Append` is a pattern that I try hard to avoid as I feel that it makes it more difficult to reason about the code when reading.

Reply:
These were precisely my concerns too, that I tried to (poorly) share in https://github.com/Azure/azure-sdk-for-net/pull/31214#discussion_r977000920

Reply:
> For my two cents, I don't mind a bit of code duplication when it is focused and serves a purpose. In this case I prefer it; I found the original version to be simpler, clearer, and have better locality of information. In particular, the reliance on mutation and side-effects in `Segments::Append` is a pattern that I try hard to avoid as I feel that it makes it more difficult to reason about the code when reading.

I think the complexity of the new approach is partly due to using a static method here. I understand that we are doing this for performance reasons, but if this were implemented as an instance method on the base class, it would be just about as simple as the original duplicated code.

Reply:
I actually thought about adding it to the base class but then realized the method would have no bearing on Message Body base because what we do there is not a general concern for all message bodies and as such it would become and even more leaky abstraction in my view. Then I thought about adding another base class for the segment based message bodies and putting it there but dismissed it due to the complexity in the inheritance chain. That's why I made it static.

Reply:
You are right - we'd probably end up needing a new class between the MessageBody and the other segment based message classes. I guess whichever approach we landed on was going to have its tradeoffs 😆 
Thanks for thinking through this with me @danielmarbach!

Reply:
Well I could have done a better job outlining my thoughts from the beginning 😂

Reply:
Hah no you're good!